torch/_expanded_weights/expanded_weights_impl.py

```diff
@@ -0,0 +1,203 @@
+    conv_kwarg_options = ['stride', 'padding', 'dilation', 'groups']
+    conv_kwarg_defaults = {'stride': 1, 'padding': 0, 'dilation': 1, 'groups': 1}
+
+    def __new__(cls, orig_weight, batch_size):
+        ret = torch.Tensor._make_subclass(cls, orig_weight.detach(), orig_weight.requires_grad)
+        if not isinstance(orig_weight, torch.Tensor):
+            raise RuntimeError(f"Can only make ExpandedWeights of Tensors, got {type(orig_weight).__name__}")
+        cls.batch_size = batch_size
+        cls.orig_weight = orig_weight
+        return ret
+
+    @classmethod
+    def __torch_function__(cls, func, _, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+        if func not in cls.handled_functions:
+            warnings.warn(f"don't have custom implementation for function {func.__name__}. Using slow fallback")
+            return SlowFallback.apply(func, *(args + tuple(kwargs.values())))
+        if func == torch.nn.functional.conv2d:
+            remaining_kwargs = 7 - len(args)
+            remaining_kwargs_options = cls.conv_kwarg_options[4 - remaining_kwargs:]
+            kwargs = {key: cls.conv_kwarg_defaults[key] for key in remaining_kwargs_options} | kwargs
+        return cls.handled_functions[func].apply(*(args + tuple(kwargs.values())))
+
+    @property
+    def shape(self):
+        return self.orig_weight.shape
+
+    @property
+    def dtype(self):
+        return self.orig_weight.dtype
+
+    @property
+    def grad(self):
+        return None
+
+    @grad.setter
+    def grad(self, value):
+        if value is None:
+            return
+        else:
+            raise RuntimeError("ExpandedWeights should never have a grad value set on it.")
+
+    @property
+    def requires_grad(self):
+        return self.orig_weight.requires_grad
+
+    @property
+    def grad_fn(self):
+        return None
+
+    def __hash__(self):
+        return id(self)
+
+    def __repr__(self):
+        return "ExpandedWeight for:\n" + self.orig_weight.__repr__() + f" with batch size {self.batch_size}"
+
+# Fallback:
+# - forward pass: run operation with de-expanded input
+# - backward pass: run autograd.grad in a for-loop to compute per-sample-grads.
+#                  This is NOT something the vmap API can handle, something more
+#                  low-level is at work here.
+class SlowFallback(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, *func_and_expanded_args):
+        func = func_and_expanded_args[0]
+        expanded_args = func_and_expanded_args[1:]
+        with torch.enable_grad():
+            output = forward_helper(func, ctx, expanded_args, 1)
+            true_output = ctx.true_outputs
+            if not isinstance(true_output, torch.Tensor):
+                raise RuntimeError(f"Fallback only works on Tensor output, got output was {type(true_output).__name__}")
+            outputs = tuple(true_output[i] for i in range(true_output.shape[0]))
+        ctx.outputs = outputs
```

Comment:
cc @albanD are there problems with this? I remember you saying something about memory leaks

It looks like we've detached the output already, so maybe this is OK

Reply:
Resolving since fallback is eliminated